design/datamover.md

```diff
@@ -0,0 +1,67 @@
+    - Image mover can specify image format type
+- End-user interaction with data mover should be exposed through the existing backup/restore operations, additionally 2 new CRDs will be introduced - DataMoverBackup
+and DataMoverRestore which would help facilitate data movement and coupling with Velero.
+
+## Non-Goals
+- This implementation will **NOT** change the existing Velero backup/restore workflow.
+  - It is possible in the future that this interface can be used to replace some legacy Velero implementations, but that is **NOT** a goal for this design/implementation.
+- This design of the interface should not include any specific details about (aside from samples):
+  - A mover type
+    - Although we are focused on volumes for phase 1, the interface should have zero knowledge of the underlying data that needs to be moved.
+    - We may provide a way to set the default data mover but still allow users to choose which data mover to user.
+  - Mover implementations
+    - Ex. How a mover handles volumes at the block vs filesystem level is not relevant to the overarching interface design
+    - On the other hand the user will set some mover specific parameters in the interface, thus the user will be aware about some implementation knowledge
+
+## Use Cases
+- Data Mover for Persistent Volume Snapshots
+  - For a given snapshot type, a capable data mover should be able to clone the snapshot data into some backup target. Of course,  we will be supporting the BSL targets but
+  the design will be extensible enough to support other kinds of backup targets as well. 
+  - The most urgent use case for this today is CSI snapshots of volumes where the given CSI driver does not guarantee durability and portability. A user creates a CSI snapshot of a volume and wishes to clone this snapshot data into some backup target. The user then would like to be able to restore a volume from this snapshot that exists in the backup target.
+- User wants to consume third party vendor data movers
+  - Example: Dell Power Protect has their own data mover to handle PVs and would like to integrate this into the native Velero backup workflow as opposed to it being a separate and disconnected process.To do this, Velero’s data mover interface has to support the ability for a vendor to bring their own data mover and have the Velero backup/restore workflow trigger its process.
+- User wants to move data associated with resources other than Persistent Volumes
+  - Example: User takes a backup of a kubevirt VM. A Kubevirt data mover would be capable of exporting the VM and pushing it to backup storage.
+  - Data mover interface should be capable of defining mover strategies for any type of kubernetes resource.
+- Data movement lifecycle will be tied up with Velero Backup/Restore.
+  - Currently, for a given backup if deleted, all the associated snapshots, restores etc. get deleted, we will follow similar behavior and also delete the moved data when backup is deleted.
```

Comment:
can we untie the lifecycle of the moved data to the lifecycle of backup CRs

We wish the backups to be durable and not dependent on any mistaken deletions etc. 
Basically providing more guarantees on availability of this data.

Reply:
+1

Reply:
@anshulahuja98 
Currently, if a backup is deleted the snapshot will be deleted, IMO deleting the moved data makes the behavior consistent with current implementation.

Reply:
Does Velero CSI plugin keeps VolumeSnapshotContent K8s API object in the original K8s cluster after a backup is complete? If it is a snapshot from EBS, you are using the snapshot which is already uploaded to S3 as a backup. In this case, it is probably fine to keep the VolumeSnapshotContent as long as you have to way to bring it back if the original K8s cluster is destroyed. If it is a local snapshot, it should not be kept after the data is uploaded to a remote backup location. In this case, it does not make sense to keep the VolumeSnapshotContent in the original K8s cluster and also the local snapshot on the storage system should be deleted after the data is uploaded.